### PR TITLE
Emit a warning when using action field from test stanza with lang dune < 1.2

### LINKED ISSUE #1471: Test stanza overwritten action should trigger an error if lang < dune 1.2
The possibility to overwrite the default `action` for a `test` stanza was released with dune `1.2` (see #1248) but using it doesn't trigger an error even if `lang` is set to an earlier version of dune.

Here I have the following `dune-project`:

```
(lang dune 1.0)
```

and the following `dune` file:

```
(test
  (name testsuite)
  (libraries
    alcotest
    terminal_size
  )
  (action
    (setenv LD_PRELOAD ./%{dep:dllmock_ioctl.so}
      (run %{test})
    )
  )
)

(rule
  (targets dllmock_ioctl.so)
  (action (run %{cc} -shared -o %{targets} %{dep:mock_ioctl.c}))
)
```

and running `dune runtest` goes just fine.

I can work on a minimal project triggering this if you need.

### PR DESCRIPTION
**Pull request by NathanReb**
Fixes #1471 

There's mostly two things going on here:

1. There should be an error when using the optional `action` field of a `test` or `tests` stanza with a syntax version < 1.2. See first commit.
2. As an error would potentially break otherwise successful builds after a dune upgrade, we'd rather just emit a warning. See second commit.

I'm not super convinced by my implementation of `2` but I thought we could discuss how to improve it here!

### CODE REVIEW

File: src/syntax.ml
Comment (emillon) [inline src/syntax.ml]: Hmm, you can't pass `fatal` to `Error.since` because of types, right? I wonder if you could instead pass the continuation to it (`Errors.warn loc` or `Errors.fail loc`).

File: src/syntax.mli
Comment (emillon) [inline src/syntax.mli]: The `fatal` argument (and its default) is probably worth documenting.

File: test/blackbox-tests/test-cases/tests-stanza-action-syntax-version/test.ml
Comment (emillon) [inline test/blackbox-tests/test-cases/tests-stanza-action-syntax-version/test.ml]: This can be an empty file:

```suggestion
```

File: src/syntax.ml
Comment (emillon) [reply src/syntax.ml]: (you'll probably need to use ksprintf)

Comment (NathanReb) [reply src/syntax.ml]: Will it really look any better?

If I understand correctly, you're suggesting to either pass `Errors.fail loc` or `fun s -> Errors.warn loc s; return ()` is that right?

If not I don't think I get what you're suggesting.

Comment (emillon) [reply src/syntax.ml]: Hmm I expected this to be shorter. Probably not much of an improvement then.

Comment (NathanReb) [reply src/syntax.ml]: Also note that `Syntax.Error.since` is exposed so when trying that out I had to make the continuation argument optional, meaning I couldn't remove the `loc` argument because it's required for the default continuation `Errors.fail loc "%s"`.

A solution might be to add some kind of `non_fatal_default: 'a` argument which would mean `fatal:false` and would specify what to return then.

Comment (NathanReb) [reply src/syntax.ml]: Just pushed a commit with the solution described above, let me know what you think!

**Comment by emillon** [comment]
I think I prefer the version without the last commit.

Can you add a changelog entry @NathanReb ?

Can we have a second look @rgrinberg ?

Thanks!

**Comment by NathanReb** [comment]
Should I remove the last commit straight away or wait for another review first?

**Comment by emillon** [comment]
We'll be able to remove it if needed, so don't worry about it :+1:

**Comment by rgrinberg** [comment]
Tbh, I'm not very convinced by `warn_and_return`. The old code might have been optically uglier (personally I don't see the difference), but at least it didn't pollute the public interface with a fairly hard to understand parameter.

**Comment by emillon** [comment]
OK! I agree! @NathanReb can you add a changelog entry (& remove the last commit while you're at it)? then i'll merge this. Thanks!

**Comment by NathanReb** [comment]
There you go!

**Review by rgrinberg — APPROVED** [review]

**Comment by emillon** [comment]
Thanks!

**Pull request merged**
